docs/RFCS/20171003_interleaved_table_joins.md

```diff
@@ -0,0 +1,779 @@
+`RowFetcher` and anything in between will need to be able to pass the index and table
+descriptors along so we can emit the rows we need, and thus `InterleavedReaderJoinerSpec`
+will need to pass this information along.
+
+#### Joining component
+
+The above covers #1 and part of #2 (ignoring the rows) of the algorithm. #2a
+is relatively straightforward.  #2b can be accomplished with a `joinerBase`
+and invoking `render` on the stored `parent` row and the streamed `child` rows.
+
+### Final output
+
+There shouldn't be any problem treating the output of the `InterleavedReaderJoiner`
+as if it was coming out of a `MergeJoiner`. In fact, the streams of the `InterleavedReaderJoiner`
+are ordered because of the primary key join. This is however not the case with
+[subset prefix joins](#3c-subset-prefix-joins).
+
+## Drawbacks
+
+Everything discussed in this PR is more efficient than the way we do joins currently.
+This is because we do one scan over the interleaved hierarchy instead of two
+separate scans for each the two table in the join.
+
+In terms of future caveats, interleaved table joins are ill-advised if the majority
+of queries are joins on a small subset of the interleaved hierarchy. In the scope of
+this RFC (where we focus on joins between two tables in the interleaved hierarchy),
+one should only have a parent and child table in a given interleaved hierarchy.
+
+Once multi-table interleaved joins are implemented (see [\[4\] Multi-table
+interleaved joins](#4-multi-table-interleaved-joins)), any scan queries
+(whether on individual tables or joins) should be on the majority of the tables
+in the interleaved hierarchy (to optimize the amount of relevant data read when
+doing [the interleaved hierarchy
+scan](#1-scanning-the-interleaved-hierarchy)).
+
+All this being said, we should be __very cautious__ prescribing usage patterns
+for interleaved tables. We currently do no support interleave schema changes
+(?) and it would be inconsistent of us if we prescribe that interleaved
+tables should only be used in parent-child relationships presently, but amend
+our prescription to include deeply-nested, hierarchial relationships once we
+implement, for instance, multi-table joins and common ancestor joins.
+
+## Rationale and Alternatives
+
+### [1] Scanning the interleaved hierarchy
+
+After discussions with @jordanlewis who has been working with indexes of interleaved
+parent and child tables, a couple of key points were brought up that dictates
+how we do the "reading" component of `InterleavedReaderJoiner`:
+1. Currently, one cannot selectively scan all parent rows without scanning
+   interleaved (grand-)\*children. Similarly, one cannot selectively scan
+   all the child rows without scanning its (grand-)\*parents and
+   (grand-)\*children.
+   - One suggestion by @jordanlewis is to maintain a secondary index that can
+     be scanned and index joined with the primary rows of the parent and/or
```

Comment:
while this is a pretty good idea I don't think we should do it automatically - a user should use create index to accomplish it. index selection can be augmented to impose a penalty for scans over parent interleaves - then such an index will get used without extra implementation work.

Reply:
I added a bullet point which emphasizes that.